bootstrap/src/basis/ordmap.ml

```diff
@@ -324,9 +324,137 @@ module T = struct
   end
 end
 include T
-include Container_common.Make_poly3_fold(T)
 include Container_array.Make_poly3_array(T)
 
+let fold_until ~init ~f t =
+  let rec fn accum f = function
+    | Empty -> accum, false
+    | Leaf {k; v} -> f accum (k, v)
+    | Node {l; k; v; n=_; h=_; r} -> begin
+        match fn accum f l with
+        | accum', true -> accum', true
+        | accum', false -> begin
+            match f accum' (k, v) with
+            | accum'', true -> accum'', true
+            | accum'', false -> fn accum'' f r
```

Comment:
```suggestion
        let accum' stop = fn accum f l in
        match stop with
        | true -> accum', true
        | false -> begin
            let accum'' stop' = f accum' (k, v) in
            match stop' with
            | true -> accum'', true
            | false -> fn accum'' f r
```
Does it matter that we're matching on something that doesn't change between the match branches? The way it is now looks fine to me, just wonder if it's better to pull the `accum` out when we're not actually matching on it.